src/nlv_object.h

```diff
@@ -5,182 +5,184 @@
 #include <assert.h>
 
 #include <memory>
-#include <nan.h>
+#include "nlv_base.h"
+
+#include "worker.h"
 using namespace node;
 using namespace v8;
 
-class NLVObjectBase;
-
-// hold a reference to a "child" object in a way that can be safely invalidated
-// if the child is destroyed by the GC before the parent.
-class NLVObjectBasePtr
-{
-public:
-  NLVObjectBasePtr(NLVObjectBase *ref) : ref_(ref), valid_(true) {}
-  bool IsValid() const { return valid_; }
-  NLVObjectBase* GetPointer() const {
-    if (!valid_) {
-      //Nan::ThrowReferenceError("attempt to access invalid NLVObjectBase pointer");
-      return NULL;
+namespace NLV {
+  NAN_INLINE unsigned int GetFlags(v8::Local<v8::Value> val) {
+    if(val->IsUndefined() || val->IsFunction()) {
+      return 0;
+    }
+    if(val->IsArray()) {
+      Local<Array> flags_ = Local<Array>::Cast(val);
+      unsigned int flags = 0;
+      for (unsigned int i = 0; i < flags_->Length(); i++)
+        flags |= flags_->Get(Nan::New<Integer>(i))->Int32Value();
+      return flags;
+    } else if(val->IsNumber()) {
+      return val->IntegerValue();
+    } else {
+      Nan::ThrowTypeError("flags must be an array or a number");
+      return 0;
     }
-
-    return ref_;
-  }
-
-  void SetInvalid() {
-    ref_ = NULL;
-    valid_ = false;
   }
 
-protected:
-  NLVObjectBase *ref_;
-  bool valid_;
-};
-
-#define NLV_STRINGIFY0(v) #v
-#define NLV_STRINGIFY(v) NLV_STRINGIFY0(v)
-
-// macro could be removed but alternate solution seems overly complex
-// https://blog.molecular-matters.com/2015/12/11/getting-the-type-of-a-template-argument-as-string-without-rtti/
-#define NLV_OBJECT_STATIC_HELPERS(class_name) \
-  static const char* ClassName() { \
-    return NLV_STRINGIFY(class_name); \
-  } \
-  friend class NLVObject;
+  template <typename ParentClass, typename HandleType, typename CleanupHandler>
+  class NLVObject : public NLVObjectBase
+  {
+  public:
+    typedef HandleType handle_type;
 
+    typedef typename std::remove_pointer<HandleType>::type HandleValue;
```

Comment:
hmm is the `std::remove_pointer` needed here? I don't actually see it being used, but it otherwise loosens the requirements for subclass by allowing you to pass in a `Type` or `Type*` - what benefit do we gain here ? (this otherwise would have most likely been programmer error)